### ISSUE TITLE
Can this action support snapshots in PRs?

### ISSUE THREAD

**Issue by joneshf-dd** [changesets/action#150]
We currently use [auto](https://intuit.github.io/auto/) for its canary release feature (each PR releases packages for testing purposes) and changesets for normal releases. But, it's not quite right. The `auto` releases don't get the correct version, nor the correct packages all the time.

It'd be ideal if we could use changeset's snapshot feature, so we thought we'd try and use that to replace `auto`. There is a pretty big issue that's preventing parity with our current setup: `auto` will comment on the PR with the published versions. Even if this action won't support snapshots in PRs with commenting the published packages, if we could at least get the actual released packages, we could piece something together. As far as I can see, there's no easy way to emulate this behavior with the changeset CLI:
- The `changeset status` command doesn't support the `--snapshot` option.
- Even if `changeset status` did support snapshots, snapshots have timestamps in them and no way to turn that off, so the version will change between the time the `changeset status` command is run and the `changeset version` command is run.
- The `changeset version` command doesn't seem to have an easy machine parseable output. It's nice and easy for a human to read, so I personally appreciate that. But, it's really hard for a machine to parse.
- The `changeset publish` command also doesn't have an easy machine parseable output.

We've tried quite a few things to get this working without resorting to re-implementing the CLI to do this task. But, it kind of seems like that might be what's necessary. Wanted to pop this question up before going down that path, since it seems like kind of a rabbit-hole.

**tl;dr;** It would be nice if this action could create snapshots for any PR that's opened, and comment back in the PR with whatever published versions.

**Comment by joneshf-dd** [comment]
> > The changeset status command doesn't support the --snapshot option.
> 
> How would you utilize this?
> 
> > Even if changeset status did support snapshots, snapshots have timestamps in them and no way to turn that off, so the version will change between the time the changeset status command is run and the changeset version command is run.
> 
> I'm not sure I understand the problem here. `changeset status` just outputs what would be released, how does the snapshot stuff relate to that? The only thing that comes to my mind is that the reported list could be smaller than the list of the packages that would actually get published at that moment.

The idea was that if we could determine what the actual version was going to be for the packages being released, we could dump that information to JSON (using the `--output=<filename>.json` option), and process it with another tool to update the PR with the released versions. But, the timestamp in the version makes that idea basically impossible. Since `changeset status` and `changeset version` would be run at different times, it's non-deterministic what the snapshot version will be.

Based on your explanation, it sounds like it just wouldn't work anyway because more packages can be released than what `changeset status` reports.

**Comment by Andarist** [comment]
According to your original TLDR this shouldn't be that big of an issue though:
> tl;dr; It would be nice if this action could create snapshots for any PR that's opened, and comment back in the PR with whatever published versions.

I mean - why do you want to call `changeset status` here? Shouldn't a script just report what already has been published rather than what would get published?

**Comment by joneshf-dd** [comment]
Oh, sorry my comment was confusing. If this action can do it, there's no other ask than that.

The stuff about `changeset status`/`changeset version`/`changeset publish` was from the attempt to implement the behavior without this action. `changeset status` is the only command that gives JSON output, so a lot of the thought went around that command. It's more a record of the things that were tried rather than an explicit ask to change any of it.

**Comment by Andarist** [comment]
> If this action can do it, there's no other ask than that.

I'm not exactly sure yet if this should be the responsibility of this action. It somewhat feels like this would have to be a "different mode" for this action or just a separate action (in this org) to avoid overloading this action too much.

However, this is just a technicality - the more interesting question here: can an action doing this be implemented using the APIs currently exposed by Changesets?

And the answer is **no**... but we are not that far away from enabling such an action to be built. If I only understand the use case correctly I believe that this is just a matter of adding that mentioned `changeset publish --json`. Then we could use the received information about released snapshot releases and post a comment using it.

**Comment by bluwy** [comment]
I'm going to close this as it looks like the discussion had happened a long time ago, but feel free to open a new one if this is still needed.
